packages/graph-retriever/src/graph_retriever/utils/top_k.py

```diff
@@ -0,0 +1,74 @@
+import heapq
+from typing import Iterable, cast
+
+from graph_retriever.content import Content
+from graph_retriever.utils.math import cosine_similarity_top_k
+
+
+def top_k(batches: Iterable[list[Content]],
+          *,
+          embedding: list[float],
+          k: int,
+          are_batches_sorted: bool = False) -> list[Content]:
+    """
+    Select the top-k contents from the given batches.
+
+    If all contents have scores, they will be used for the comparison rather
+    than being computed.
+
+    Parameters
+    ----------
+    batches : Iterable[list[Content]]
+        The batches of content to select the top-K from.
+    embedding: list[float]
+        The embedding we're looking for.
+    k : int
+        The number of items to select.
+    are_batches_sorted : bool, default False
+        Whether the content of each batch are sorted. If true, and all content
+        has scores, a more efficient top-K selection will be used which doesn't
+        need to consider all of each batch.
+    """
+
+    # TODO: Consider passing threshold here to limit results.
+
+    if all(c.score is not None for batch in batches for c in batch):
+        sorted_items: Iterable[Content]
+        if are_batches_sorted:
+            sorted_items = heapq.merge(*batches, key=_score, reverse=True)
```

Comment:
This case (all items have scores and batches are sorted) means we don't need to even look at all of each list. I'm not sure how often it would arise though, and it means that each adapter needs to indicate that the results are sorted.

The most unlikely case is going to be `get` / `aget` -- currently, I don't think most stores even consider the embedding in their implemnetation of that, so they won't return a score nor sort by it. It is *possible* that at least Astra (and maybe some others) could do that, which would allow this path to be used. Thoughts -- should I take steps to make it possible to hit this path and use it?

Reply:
maybe this could be a future simplification in some cases?